### DISCUSSION TITLE
Velocity support

### DISCUSSION THREAD

**Discussion by natr566** [Zrips/CMI#2515] [Feature Requests]
**Description of feature/improvement request:**
Add support for velocity
---
### Detailed explanation of what it should do, when and in what situation: ###
Just like the bungee addon but for Velocity
https://www.velocitypowered.com/
### Basic explanation as to why this feature is needed and/or useful to have: ###
Because velocity is cleaner then bungee and it's great.

**Comment by mrfloris** [comment]
I see more people use velocity as they switch to paper, so makes sense to consider supporting it if it doesn't already.

**Comment by dyanakiev** [comment]
Bump?

**Comment by Tarlak333** [comment]
Yes, please add velocity support.  I just migrated my self-hosted multi-server network over to Velocity a month ago and am desperately seeking quality plugins that can be used across all the back-end servers.  (Network wide chat channels, player nameplates/tags, broadcast messaging, and a network playtime rewards (credits) system are some features I'm seeking.)  I'm currently using Anvil/Catalyst to fill some of this need but it's VERY limited.

Also: It would be useful to to have the option to have some data into a "network" (shared) MySQL table, in addition to the "per-server" tables.

Thank you for the great work you do for the MC community,

Tarlak333

**Comment by Albert481** [comment]
+1

**Comment by Parthurax** [comment]
+1

**Comment by Euphillya** [comment]
+1

**Comment by AppleMacOS** [comment]
Velocity support would be extremely nice to have, since I've been looking into switching to Velocity but the only thing holding me back right now is CMI support for Velocity.

**Comment by Stixil** [comment]
Bump

**Comment by MonzterDev** [comment]
Just swapped to Velocity, sucks I have to drop CMI :(

Please add support!

**Comment by ghost** [comment]
+1

**Comment by myavuzokumus** [comment]
Perfectly working with snap. Can use it until the update comes.

**Reply by iDutch** [reply]
Can confirm this works. Thank you for this tip. Now I can ditch Advanced Portals again :P

**Reply by JamesBach93** [reply]
Strange, i use snap too. Actually only for "bungeejoinmessage". But want CMI on this too. After placing the Plugin in the Folder and restart the Proxy i dont get an CMI or CMIB Folder. On /cmi version there is no bungee info. And Bungee Placeholder dont work. https://i.imgur.com/KyqFR1Y.png

**Reply by mrfloris** [reply]
> Strange, i use snap too. Actually only for "bungeejoinmessage". But want CMI on this too. After placing the Plugin in the Folder and restart the Proxy i dont get an CMI or CMIB Folder. On /cmi version there is no bungee info. And Bungee Placeholder dont work. https://i.imgur.com/KyqFR1Y.png

CMI goes to the servers
CMIB jar goes to bungee
i dont know about velocity and snap, as cmi doesn't have support for this. 
CMI doesn't go to the proxy unless velocity/snap installation instructions for cmi demand this. which i am not aware of.

**Reply by iDutch** [reply]
> > Strange, i use snap too. Actually only for "bungeejoinmessage". But want CMI on this too. After placing the Plugin in the Folder and restart the Proxy i dont get an CMI or CMIB Folder. On /cmi version there is no bungee info. And Bungee Placeholder dont work. https://i.imgur.com/KyqFR1Y.png
> 
> CMI goes to the servers CMIB jar goes to bungee i dont know about velocity and snap, as cmi doesn't have support for this. CMI doesn't go to the proxy unless velocity/snap installation instructions for cmi demand this. which i am not aware of.

You just need the CMIB jar in the plugin directory inside the snap directory.

**Comment by Felleus** [comment]
+1

**Comment by AZFire1** [comment]
+1

**Comment by Nil0375** [comment]
+1

**Comment by nhanledev** [comment]
Paper plans to drop Waterfall, please support Velocity asap

**Comment by Baktus79** [comment]
Any news about this?

**Comment by ElTodoFull** [comment]
Velocity should be a standard by now. As soon as it gets Velocity support, I will buy CMI ASAP.
Thank you in advance!

**Comment by mrfloris** [comment]
The question for proxy support is pretty high, and understandably so. For existing and new buyers. MySQL user-data bridge plugin can help those until there is some support. If zrips decides to add it. I personally have no info if it's in the making or still in the consideration stage. It's quite an undertaking for a big plugin like this and from what i understand not something he can gradually introduce per feature.

**Comment by prototype464** [comment]
Please add Velocity support, I desperately need this feature for my network, and Paper is planning on dropping Waterfall. BungeeCord is ancient!

**Comment by ghost** [comment]
+1

**Comment by JamesBach93** [comment]
+1

**Comment by 7hatGuyAJ** [comment]
"I see more people use velocity as they switch to paper, so makes sense to consider supporting it if it doesn't already." [mrfloris]
 [on 5 Oct 2020].

Unfortunately Unless Support is worked on for velocity soon I will have to drop CMI for another plugin, This saddens me but unfortunately we have to evolve and move on.

**Comment by NoltoxGit** [comment]
Please make CMIBungee support Velocity plz.

**Comment by LucidAPs** [comment]
Please Velocity support.

**Comment by Fabrimat** [comment]
Almost 3 years and still nothing?

**Comment by nhanledev** [comment]
If only he open source the CMI Bungee module and let us hook into CMI for that feature, it is no hope because CMI heavily utilizes bungeecord's channel

**Comment by mrfloris** [comment]
I think it's fair to say that velocity support is the most requested feature, hopefully we will gradually see more network-data-sync features.

**Comment by izimen** [comment]
Good idea to add velocity support, velocity is much better software than bungeecord, I'm for it

**Comment by Baktus79** [comment]
I'm willing to pay @Zrips to do Velocity support, if it increases motivation. Name your price and Im sure all the others around here would be willing to do the same. Or make it premium.

**Reply by CloudeLecaw** [reply]
cmi is allready Premium .. ;)

**Reply by Baktus79** [reply]
An yet no Velocity support

**Reply by Baktus79** [reply]
Not sure why you thought that was a good comment

**Reply by nhanledev** [reply]
And why do you expect velocity support when even bungee has not stable yet? lol

**Reply by Baktus79** [reply]
Because Velocity is better?

**Reply by Baktus79** [reply]
lol

**Reply by nhanledev** [reply]
I hope the bungeecord module gets better before it is ported into velocity lol.

**Reply by Baktus79** [reply]
I have never had any problems with the Bungeecord module. Let me correct you a little bit. I never said I expect Velocity support, this is a wish on my part. That you don't like Velocity is one thing, but I think the users of CMI have had their say, Velocity support is one of the most requested features for CMI. I think it speaks for itself.

**Reply by nhanledev** [reply]
Haha please don't mind me I am just trying to ping this issue as much as possible, and hope zrips can spend some of his time to notice this

**Reply by mrfloris** [reply]
> Haha please don't mind me I am just trying to ping this issue as much as possible, and hope zrips can spend some of his time to notice this

Do not uselessly spam and ping conversations and issues. It does NOT result in things being sped up. It will only result in you getting blocked from participating. 

Contribute relevant information. Dont spam for attention

**Comment by mrfloris** [comment]
From Discord; `@harold#1385` > Sync playtime across servers.

**Comment by JesterRomut** [comment]
+1

**Comment by Turbovadim** [comment]
+1

**Comment by sectlying** [comment]
+1

**Comment by CormacZ** [comment]
+1

**Comment by SnonSS** [comment]
+

**Comment by SolarPepper** [comment]
+1

**Comment by TomLewis** [comment]
I came here to request Velocity support, and noticed this post was already here! So treat this comment as another request for velocity!

**Comment by ghost** [comment]
+1
Please add velocity support.. we've been waiting 4 years soon..

**Comment by Vort3chs** [comment]
Velocity support would be a game changer for me.

+1

**Comment by QueeZizz** [comment]
+1

**Comment by CloudeLecaw** [comment]
Does it still worth to think about Velocity support? i read somewhere there is this new thing -> folia? or is this a complete different server fork?

**Reply by Baktus79** [reply]
Folia is something totally different, Velocity is used for proxy. The Paper team is creating Folia to be a multithreaded Paper fork.

**Reply by CloudeLecaw** [reply]
thanks for pointing this out. well i dont have much to do with all this..we just use a normal paper one.

**Comment by compieter-gh** [comment]
+1

**Comment by DarkKingArthas** [comment]
+1
Velocity is becoming more and more popular. Safer and most as I can see are switching to it. CMI is the only plugin in my collection that does not support velocity at the moment.

**Comment by natr566** [comment]
I guess discussions just sit round and do nothing
(Well its only been 3 years, 7 months since i asked about this)

**Comment by CloudeLecaw** [comment]
never mind, time for the next 3 years and seven months to wait for ;)

**Comment by compieter-gh** [comment]
Kinda why i am ditching CMI as i feel like velocity & bungeecord support for that fact both just arent that great
I went back too Essentialsx with other open-source plugins too fill in the gaps.

**Comment by kougou888** [comment]
+1

**Comment by pintux98** [comment]
Hello everyone, I made a Velocity port of the bungeecord plugin.
I have tested it a bit, it seems to work fine for me.

If you need , you can download from spigot: 
https://www.spigotmc.org/resources/cmi-velocity.111751/

**Reply by LucidAPs** [reply]
If this works correctly I can finally switch to velocity!

**Reply by pintux98** [reply]
I am using this, and for me so far is working fine!

**Reply by kougou888** [reply]
Thank you for your contribution, but I have some problems when using it:

[00:54:05 ERROR]: Couldn't pass DisconnectEvent to cmiv
java.lang.IllegalStateException: Not connected to server!
        at com.velocitypowered.proxy.connection.backend.VelocityServerConnection.ensureConnected(VelocityServerConnection.java:218) ~[velocity-3.2.0-SNAPSHOT-260.jar:3.2.0-SNAPSHOT (git-37fedf50-b260)]
        at com.velocitypowered.proxy.connection.backend.VelocityServerConnection.sendPluginMessage(VelocityServerConnection.java:276) ~[velocity-3.2.0-SNAPSHOT-260.jar:3.2.0-SNAPSHOT (git-37fedf50-b260)]
        at com.velocitypowered.proxy.connection.backend.VelocityServerConnection.sendPluginMessage(VelocityServerConnection.java:262) ~[velocity-3.2.0-SNAPSHOT-260.jar:3.2.0-SNAPSHOT (git-37fedf50-b260)]
        at it.pintux.CMIV.sendServerEvent(CMIV.java:122) ~[?:?]
        at it.pintux.CMIV.onDisconnect(CMIV.java:107) ~[?:?]
        at it.pintux.Lmbda$4.execute(Unknown Source) ~[?:?]
        at com.velocitypowered.proxy.event.UntargetedEventHandler$VoidHandler.lambda$buildHandler$0(UntargetedEventHandler.java:56) ~[velocity-3.2.0-SNAPSHOT-260.jar:3.2.0-SNAPSHOT (git-37fedf50-b260)]
        at com.velocitypowered.proxy.event.VelocityEventManager.fire(VelocityEventManager.java:597) ~[velocity-3.2.0-SNAPSHOT-260.jar:3.2.0-SNAPSHOT (git-37fedf50-b260)]
        at com.velocitypowered.proxy.event.VelocityEventManager.lambda$fire$5(VelocityEventManager.java:478) ~[velocity-3.2.0-SNAPSHOT-260.jar:3.2.0-SNAPSHOT (git-37fedf50-b260)]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]

**Reply by pintux98** [reply]
Hi, i’ll check it out and fix in a few mins. Keep an eye on the spigot page

**Reply by pintux98** [reply]
Hi @kougou888 I just posted an update

**Reply by TomLewis** [reply]
Will you be putting the code up on github so we can also have an issue tracker?

**Reply by pintux98** [reply]
https://github.com/pintux98/CMIV

**Comment by Baktus79** [comment]
Now it's official!
> While Waterfall received support for 1.20.2 pretty early on as part of BungeeCord upstream updates, its support was pretty broken for the first few days and weeks after the release and still does not properly handle the new protocol changes in some places. In general, Waterfall is unlikely to receive our full attention given that Velocity is meant to be its more performant, stable, and secure successor. In similar fashion to us retiring Travertine a while ago, the same will happen to Waterfall in the future. For now though, we will continue providing you with upstream updates at the very least.

[https://forums.papermc.io/threads/paper-velocity-1-20-2.920/](https://forums.papermc.io/threads/paper-velocity-1-20-2.920/)

**Reply by TomLewis** [reply]
Hence why everyone here wants support haha so many people alraedy moved to velocity!

**Comment by mrfloris** [comment]
https://www.spigotmc.org/resources/cmi-298-commands-insane-kits-portals-essentials-economy-mysql-sqlite-much-more.3742/update?update=521168

There is a mention in the changelog for velocity support in today's CMI release.
You can get the free cmiv jar file here: https://www.zrips.net/cmiv/

**Comment by TomLewis** [comment]
Fantastic news!
